TeleSharp.Generator/Program.cs

```diff
@@ -16,39 +17,195 @@ class Program
+                Console.WriteLine("For more information, please read the manual,");
+                Console.WriteLine("or visit the GitHub page.");
+                Console.WriteLine("Submit your bug reports to our GitHub repository.");
+            }
+            else
+            {
+                Console.WriteLine("Try `TeleSharp.Generator --help' for more options.");
+            }
+        }
+
+        enum Format
+        {
+            TL,
+            JSON
+        }
+
         static void Main(string[] args)
         {
+            string Json = "";
+            string inputPath = "";
+            string outputPath = "";
+            Format format = Format.TL;
+            bool forceFormat = false;
+            bool outputJson = false;
+            bool showHelp = false;
+
 
             string AbsStyle = File.ReadAllText("ConstructorAbs.tmp");
             string NormalStyle = File.ReadAllText("Constructor.tmp");
             string MethodStyle = File.ReadAllText("Method.tmp");
-            //string method = File.ReadAllText("constructor.tt");
-            string Json = "";
-            string url;
-            if (args.Count() == 0) url = "tl-schema.json"; else url = args[0];
+            string TargetNamespace = "TeleSharp.TL";
+
+            bool invalidFormat = false;
+            bool invalidTargetNamespace = false;
+
+            OptionSet optionset = new OptionSet()
+                .Add("h|help", h => showHelp = h != null)
+                .Add("f|format=", f =>
+                {
+                    f = f ?? "";
+                    if (f != null)
+                        forceFormat = true;
+                    switch (f.ToLower())
+                    {
+                        case "tl":
+                            format = Format.TL;
+                            break;
+                        case "json":
+                            format = Format.JSON;
+                            break;
+                        case "":
+                            format = Format.TL;
+                            break;
+                        default:
+                            invalidFormat = true;
+                            break;
+                    }
+                })
+                .Add("target-namespace=", a =>
+                {
+                    if (!string.IsNullOrEmpty(a))
+                    {
+                        Match m = Regex.Match(a, @"(@?[a-z_A-Z]\w+(?:\.@?[a-z_A-Z]\w+)*)");
+                        if (m.Success)
+                        {
+                            TargetNamespace = m.Groups[0].Value;
+                        }
+                    }
+                    else
+                    {
+                        invalidTargetNamespace = true;
+                    }
+                })
+                .Add("output-json", a => outputJson = a != null)
+                .Add("template-abstract=", a => AbsStyle = (a != null) ? File.ReadAllText(a) : AbsStyle)
```

Comment:
A good thing I added was the custom template. I think it will be usable to someone who wants to develop his/her own client library.